processing/src/test/java/org/apache/druid/segment/join/PostJoinCursorTest.java

```diff
@@ -0,0 +1,249 @@
+  }
+
+  /**
+   * Simulates infinite segment by using a base cursor with advance() and advanceInterruptibly()
+   * reduced to a no-op.
+   */
+  private static class TestInfiniteQueryableIndexSegment extends QueryableIndexSegment
+  {
+
+    private final StorageAdapter testStorageAdaptor;
+
+    public TestInfiniteQueryableIndexSegment(QueryableIndex index, SegmentId segmentId)
+    {
+      super(index, segmentId);
+      testStorageAdaptor = new InfiniteQueryableIndexStorageAdapter(index);
+    }
+
+    @Override
+    public StorageAdapter asStorageAdapter()
+    {
+      return testStorageAdaptor;
+    }
+  }
+
+  private static class InfiniteQueryableIndexStorageAdapter extends QueryableIndexStorageAdapter
+  {
+    public InfiniteQueryableIndexStorageAdapter(QueryableIndex index)
+    {
+      super(index);
+    }
+
+    @Override
+    public Sequence<Cursor> makeCursors(
+        @Nullable Filter filter,
+        Interval interval,
+        VirtualColumns virtualColumns,
+        Granularity gran,
+        boolean descending,
+        @Nullable QueryMetrics<?> queryMetrics
+    )
+    {
+      return super.makeCursors(filter, interval, virtualColumns, gran, descending, queryMetrics).map(
+          CursorNoAdvance::new);
+    }
+
+    private static class CursorNoAdvance implements Cursor
+    {
+      Cursor cursor;
+
+      public CursorNoAdvance(Cursor cursor)
+      {
+        this.cursor = cursor;
+      }
+
+      @Override
+      public ColumnSelectorFactory getColumnSelectorFactory()
+      {
+        return cursor.getColumnSelectorFactory();
+      }
+
+      @Override
+      public DateTime getTime()
+      {
+        return cursor.getTime();
+      }
+
+      @Override
+      public void advance()
+      {
+        // Do nothing to simulate infinite rows
+      }
+
+      @Override
+      public void advanceUninterruptibly()
+      {
+        // Do nothing to simulate infinite rows
+      }
+
+      @Override
+      public boolean isDone()
+      {
+        return cursor.isDone();
+      }
+
+      @Override
+      public boolean isDoneOrInterrupted()
+      {
+        return cursor.isDoneOrInterrupted();
+      }
+
+      @Override
+      public void reset()
+      {
+
+      }
+    }
+  }
+
+  private static class ExceptionHandler implements Thread.UncaughtExceptionHandler
+  {
+
+    Throwable exception;
+
+    @Override
+    public void uncaughtException(Thread t, Throwable e)
+    {
+      exception = e;
+    }
+
+    public Throwable getException()
+    {
+      return exception;
+    }
+  }
+
+  @Test
+  public void testAdvanceWithInterruption()
+  {
+    Thread joinCursorThread = new Thread(() -> makeCursorAndAdvance());
+    ExceptionHandler exceptionHandler = new ExceptionHandler();
+    joinCursorThread.setUncaughtExceptionHandler(exceptionHandler);
+    joinCursorThread.start();
+    try {
+      sleep(5000);
+      joinCursorThread.interrupt();
+      sleep(5000);
```

Comment:
A 10 second test is expensive. Rather, we should use a count latch or something to wait for X records to be processed. The latch wait can have 5 second timeout. Similarly, the 5 second sleep after interrupting should be removed and we can wait in a loop or use count latch again in ExceptionHandler. 

I also like https://github.com/awaitility/awaitility/wiki/Usage#simple in case you want to bring that in here.

Reply:
Went with a latch, but thanks for the pointer to this library.